File: sdk/typescript/src/api/bluetooth.ts
import type { HttpTransport } from '../transport.js';
import type { BluetoothDevice } from '../types.js';

const BLUETOOTH_DEVICE_FIELDS = /* GraphQL */ `
  fragment BluetoothDeviceFields on BluetoothDevice {
    address name paired trusted connected rssi
  }
`;

export class BluetoothApi {
  constructor(private readonly http: HttpTransport) {}

  /** List paired/known Bluetooth devices (Linux only) */
  async devices(): Promise<BluetoothDevice[]> {
    const data = await this.http.execute<{ bluetoothDevices: BluetoothDevice[] }>(/* GraphQL */ `
      ${BLUETOOTH_DEVICE_FIELDS}
      query BluetoothDevices { bluetoothDevices { ...BluetoothDeviceFields } }
    `);
    return data.bluetoothDevices;
  }

  /** Scan for nearby Bluetooth devices (Linux only) */
  async scan(timeoutSecs?: number): Promise<BluetoothDevice[]> {
    const data = await this.http.execute<{ bluetoothScan: BluetoothDevice[] }>(/* GraphQL */ `
      ${BLUETOOTH_DEVICE_FIELDS}
      mutation BluetoothScan($timeoutSecs: Int) {
        bluetoothScan(timeoutSecs: $timeoutSecs) { ...BluetoothDeviceFields }
      }
    `, { timeoutSecs });
    return data.bluetoothScan;
  }

  /** Connect to a Bluetooth device by address (Linux only) */
  async connect(address: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation BluetoothConnect($address: String!) { bluetoothConnect(address: $address) }
    `, { address });
  }

  /** Disconnect a Bluetooth device by address (Linux only) */
  async disconnect(address: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation BluetoothDisconnect($address: String!) { bluetoothDisconnect(address: $address) }
    `, { address });
  }
}

File: sdk/ruby/lib/rockbox/api/bluetooth.rb
# frozen_string_literal: true

require_relative "../types"

module Rockbox
  module Api
    class Bluetooth
      FIELDS = <<~GQL
        fragment BluetoothDeviceFields on BluetoothDevice {
          address name paired trusted connected rssi
        }
      GQL

      def initialize(http)
        @http = http
      end

      # List paired/known Bluetooth devices (Linux only).
      def devices
        data = @http.execute("#{FIELDS}\nquery BluetoothDevices { bluetoothDevices { ...BluetoothDeviceFields } }")
        Array(data[:bluetooth_devices]).map { |d| BluetoothDevice.from_hash(d) }
      end

      # Scan for nearby devices (Linux only).
      def scan(timeout: nil)
        data = @http.execute(<<~GQL, timeout ? { timeout_secs: timeout } : nil)
          #{FIELDS}
          mutation BluetoothScan($timeoutSecs: Int) {
            bluetoothScan(timeoutSecs: $timeoutSecs) { ...BluetoothDeviceFields }
          }
        GQL
        Array(data[:bluetooth_scan]).map { |d| BluetoothDevice.from_hash(d) }
      end

      def connect(address)
        @http.execute(
          "mutation BluetoothConnect($address: String!) { bluetoothConnect(address: $address) }",
          { address: address }
        )
        nil
      end

      def disconnect(address)
        @http.execute(
          "mutation BluetoothDisconnect($address: String!) { bluetoothDisconnect(address: $address) }",
          { address: address }
        )
        nil
      end
    end
  end
end
